src/Constraint/Constraint.php

```diff
@@ -179,6 +182,10 @@ public function versionCompare($a, $b, $operator, $compareBranches = false)
      */
     public function matchSpecific(Constraint $provider, $compareBranches = false)
     {
+        $cacheKey = $this->operator.(int)$compareBranches.$provider->version;
```

Comment:
Here too I'd assume `spl_object_id($provider).$compareBranches` would be a better cacheKey?